compositor_pipeline/src/pipeline/input/whip.rs

```diff
@@ -0,0 +1,286 @@
+pub enum WhipReceiverError {
+    #[error("Failed to add input {0} to input_connections hashmap")]
+    WhipWhepStateAddInput(InputId),
+}
+
+#[derive(Debug, Clone)]
+pub struct WhipReceiverOptions {
+    pub video: Option<InputVideoStream>,
+    pub audio: Option<InputAudioStream>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InputVideoStream {
+    pub options: decoder::VideoDecoderOptions,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InputAudioStream {
+    pub options: decoder::OpusDecoderOptions,
+}
+
+pub struct WhipReceiver {
+    whip_whep_state: Arc<WhipWhepState>,
+    input_id: InputId,
+}
+
+impl WhipReceiver {
+    pub(super) fn start_new_input(
+        input_id: &InputId,
+        opts: WhipReceiverOptions,
+        pipeline_ctx: &PipelineCtx,
+    ) -> Result<InputInitResult, WhipReceiverError> {
+        let bearer_token = generate_token();
+        let whip_whep_state = pipeline_ctx.whip_whep_state.clone();
+
+        let (mut video_tx, mut video_rx) = (None, None);
+        let (mut video_tx_crossbeam, mut video_rx_crossbeam) = (None, None);
+
+        if opts.video.is_some() {
+            let (tx, rx) = mpsc::channel(100);
+            video_tx = Some(tx);
+            video_rx = Some(rx);
+            let (tx_crossbeam, rx_crossbeam) = crossbeam_channel::bounded(100);
+            video_tx_crossbeam = Some(tx_crossbeam);
+            video_rx_crossbeam = Some(rx_crossbeam);
+        }
+
+        let (mut audio_tx, mut audio_rx) = (None, None);
+        let (mut audio_tx_crossbeam, mut audio_rx_crossbeam) = (None, None);
+
+        if opts.audio.is_some() {
+            let (tx, rx) = mpsc::channel(100);
+            audio_tx = Some(tx);
+            audio_rx = Some(rx);
+            let (tx_crossbeam, rx_crossbeam) = crossbeam_channel::bounded(100);
+            audio_tx_crossbeam = Some(tx_crossbeam);
+            audio_rx_crossbeam = Some(rx_crossbeam);
+        }
+
+        let depayloader = Arc::from(Mutex::new(Depayloader::new(&opts)));
+
+        Self::start_forwarding_thread(
+            input_id,
+            video_rx,
+            audio_rx,
+            video_tx_crossbeam,
+            audio_tx_crossbeam,
+        );
+
+        if let Ok(mut input_connections) = whip_whep_state.input_connections.lock() {
+            input_connections.insert(
+                input_id.clone(),
+                WhipInputConnectionOptions {
+                    audio_sender: audio_tx.clone(),
+                    video_sender: video_tx.clone(),
+                    bearer_token: Some(bearer_token.clone()),
+                    peer_connection: None,
+                    start_time_vid: None,
+                    start_time_aud: None,
+                    depayloader,
+                },
+            );
+        } else {
+            return Err(WhipReceiverError::WhipWhepStateAddInput(input_id.clone()));
+        }
+
+        let video = match (video_rx_crossbeam, opts.video) {
```

Comment:
you are checking video in bunch of places, in general you want to only check it in one place and do everything there.

When you are creating video_rx_crossbeam you know on type level that it is not None, but here you need to handle again which is not necessary.

Separation like this might be good if those places become to complicated, but in most cases there are better ways (type safe) that can solve those situation